### PR TITLE
Print feature Proof-of-Concept

### PR DESCRIPTION
**Pull request by tomsib2001**

### CODE REVIEW

**Comment by tomsib2001** [comment]
Here is the test file I submitted during the presentation:

    (* Printing stuff in Coq for the benefit of debugging  *)
    (* and some fun with terminal recursivity              *) 

    Check print.
    Check print_id.

    Print id.

    Declare Printing Effect id.

    Eval compute in id 3.

    Declare Printing Effect id.

    Eval cbv in (print (3 + 3)).
    Eval cbn in print (3 + 3).

    Fixpoint factorial (n : nat) := 
      match n with
        | O => 1
        | S k => print_id (n * factorial k)
      end.

    Eval compute in factorial 6.

    Print print.
    Print print_id.

    Fixpoint factorial2 (n : nat) := 
      match n with
        | O => 1
        | S k => let res := (n * factorial k) in 
                 let _ := print res in
                 res
      end.

    Eval compute in factorial2 6.

    Eval cbn in factorial 6.
    Eval cbn in factorial2 6.


     Eval compute in print (print).

**Comment by herbelin** [comment]
Are there objections to make print Polymorphic so that "print print" could type-check?

The convention that the 2nd argument is printed is somehow arbitrary and it should be made more flexible eventually.

**Comment by tomsib2001** [comment]
No objection here. The fact that it is the second argument is not only arbitrary, but it actually breaks in some cases (for some reason I can't find it right now but I'll post an example as soon as I find it)

**Comment by maximedenes** [comment]
test build

**Comment by tomsib2001** [comment]
test build

**Comment by maximedenes** [comment]
Ok, something I realize only now is that this PR doesn't actually provide a hook for a printing function. The only thing that one can register using the Print_hook module is the *name* of the constant representing the print function.

I am not sure if this is compatible with what we put on the roadmap (keeping only the hooks).

**Comment by herbelin** [comment]
I'm very sorry but I must confess that I don't understand, and a fortiori then did not understand what was the result of the discussion at the roadmap session. Can someone remind me what was expected to be changed. Thanks in advance.

**Comment by maximedenes** [comment]
The conclusion of the roadmap discussion was: we integrate only the hook, to make it possible to later provide a printing function through a plugin. I could not apply this decision because no such hook is provided by this PR. The confusion comes from the ill-named Print_hook.ml file.

**Comment by maximedenes** [comment]
Also, now that I understand the code of CBV, I don't see why we need a reify_value function.

**Comment by maximedenes** [comment]
So what needs to be done is to implement the minimal machinery to be integrated on Coq's side, to make it possible to experiment this proof-of-concept feature from a plugin.

**Comment by herbelin** [comment]
@maximedenes: Do you mean that the result of the discussion was 1) to extend the printing effect into an arbitrary effect (of type "constr -> unit"?) together, say, with a number, telling on which argument to apply the effect? 2) To offer the syntax as a plugin, say plugins/reductioneffect? Is that correct?

cbv: the machine is working on a specific datatype "cbv_value". How can you print a "cbv_value" without first translating it to "constr"?

File: pretyping/cbv.ml [hunk 1]
```diff
@@ -325,7 +375,7 @@ and cbv_stack_value info env = function
  * normalize (even under binders) the applied terms and we build the
  * final term
  *)
-let rec apply_stack info t = function
+and apply_stack info t = function
   | TOP -> t
```
Comment (JasonGross) [inline pretyping/cbv.ml]: What's the reason for this to become mutually recursive?

Comment (herbelin) [inline pretyping/cbv.ml]: Probably the unnecessary residual of a mutual recursivity whose need finally disappeared...

Comment (JasonGross) [inline pretyping/cbv.ml]: As in, left over from a merge conflict?  This PR seems to be turning a previously non-mutually recursive function into a mutually recursive one.

**Comment by herbelin** [comment]
Rebase against trunk available at https://github.com/herbelin/github-coq, branch print_feature_poc.

**Comment by maximedenes** [comment]
@herbelin cbv_norm_value already reifies cbv_value to constr. Simply call it with the CClosure.nored flag.

**Comment by herbelin** [comment]
Ah, good idea. We did not think at using it with the nored flag.

**Comment by herbelin** [comment]
Hi all,

I came back to think at this PR. We could not really talk about it at the WG but my understanding of the situation is the following:
- Using `cbv_norm_value` seems reasonable for reification; it would do quite a bit of useless job (like turning terms into sequent calculus expressions <term|stack> back and forth in `cbv_norm_term` while the `constr` is already there without having to be inspected), but the gain in code factorization is interesting. Don't know the opinion of others here.
- Pluginization: here is my understanding of the discussion, leading to a generalized model:
  - `print_hook.ml` renamed into `redeffect_hook.ml` (in `pretyping`); generalize `set_print_ref : global_reference -> unit` into `set_effect_ref : global_reference -> (env -> constr -> unit) -> unit`
  - new directory `plugins/redeffect`
  - move `extravernac.ml4` renamed into something like `redeffect.ml4` there (this provides the `VERNAC EXTEND` "Declare Printing Effect"; eventually have a stronger check that the constant used to trigger the printing effect has type `forall A:Type, A -> A` for some universe)
  - in particular, the printing effect is declared by the plugin, not directly in `redeffect_hook.ml`

This seems to implement what was discussed long time ago, right?

Now, we can eventually provide other effects or variants of printing effects, such as e.g. (naively):
- "print_and_stop_expecting_a_key_pressed" (with command say `Declare Printing Effect ref (interactive)`, or with any other way to give an attribute)
- "read_term": one such "read_term" could also be triggered say by a constant of type `forall A, A -> A` with the following effect: if the term of type `A` is an abstraction `fun x:T => M`, then read a term `N` interactively, check it has type `T` and returns the term `fun _:T => (fun x:T => M) N` (but this would require to change the type of the effect hook from `set_effect : global_reference -> (constr -> unit) -> unit` to say `set_effect : global_reference -> (env -> constr -> constr option) -> unit`
- etc.

Of course, users could also themselves implement plugins defining effects

**Comment by herbelin** [comment]
Hi, @tomsib2001 being busy, I made an update of this feature.

I renounced to reuse `cbv_norm_value` in `cbv` (passing a `no_red` flag and using a forward definition was making the code more complicated than necessary).

I implemented a parametrization of the effect in a plugin. Currently, there is only the printing effect which is activated by importing the module `PrintingEffect`.

I extended the support of the effect to `simpl`, `hnf`, ...

There is a little theoretical overhead which I did not measure.

Somehow, the effect shows strange behaviors of these reduction functions. To be investigated further...

The branch is [here](https://github.com/herbelin/github-coq/tree/print_feature_poc2).

**Pull request closed without merging**